designer/live-data-collector/src/main/scala/pl/touk/nussknacker/engine/livedata/SlidingWindowCounter.scala

```diff
@@ -0,0 +1,60 @@
+package pl.touk.nussknacker.engine.livedata
+
+import java.time.{Clock, Instant}
+import java.util.concurrent.{ConcurrentHashMap, ConcurrentLinkedQueue}
+import java.util.concurrent.atomic.AtomicLong
+import scala.jdk.CollectionConverters._
+
+private[livedata] class SlidingWindowCounter[T](
+    counterCreatedAt: Instant,
+    windowSizeSeconds: Int
+)(implicit clock: Clock) {
+
+  private val buckets = new ConcurrentHashMap[Long, ConcurrentLinkedQueue[T]]()
+
+  def add(event: T): Unit = {
+    val currentEpochSecond = now()
+    cleanOldBuckets(currentEpochSecond)
+    val bucket = buckets.computeIfAbsent(currentEpochSecond, _ => new ConcurrentLinkedQueue[T]())
+    bucket.add(event)
+  }
+
+  def getThroughput: Map[T, BigDecimal] = {
+    val currentEpochSecond = now()
+    cleanOldBuckets(currentEpochSecond)
+
+    // We want to calculate correct throughput just after the scenario is started
+    val windowStart      = Math.max(counterCreatedAt.getEpochSecond, currentEpochSecond - windowSizeSeconds)
+    val windowsEnd       = currentEpochSecond
+    val samplingInterval = windowsEnd - windowStart + 1
+
+    buckets.asScala.values
+      .flatMap(_.asScala)
+      .toList
+      .groupBy(identity)
+      .view
+      .map { case (key, value) => (key, value.size) }
+      .toMap
+      .map { case (transition, count) =>
+        transition -> BigDecimal(count)./(samplingInterval).setScale(4, BigDecimal.RoundingMode.HALF_UP)
+      }
+  }
+
+  private val lastCleaned = new AtomicLong(0)
+
+  private def cleanOldBuckets(now: Long): Unit = synchronized {
+    // Clean old buckets at most once per second, not on each call
+    if (lastCleaned.getAndSet(now) != now) {
```

Comment:
For `windowSizeSeconds` around hours or day it will slow down events processing due to O(n) complexity. For similar problem (sliding window aggregations` we used `SortedMap` which has O(log n) complexity, see: `AggregatorFunction` and `FlinkRangeMap`. WDYT about this approach? Should we apply it here as well?

Reply:
I changed `ConcurrentSkipListMap` to `ConcurrentSkipListMap`, which according to Java docs provides log(n) time cost for the get, put and remove.

Reply:
The comment is attached to `cleanOldBuckets` where you filter set by date. Take a look at `SortedMap.rangeTo` characteristics.

Reply:
Ok, so:
- I already changed the map implementation to ConcurrentSkipListMap
- now I also changed the cleaning implementation to use `.headMap(cutoff)`, which has `O(log n)` complexity
- this cleaning is invoked at most once per second - there is a condition that prevents more than one cleanup per second, irrespective of number of events per second `if (lastCleaned.getAndSet(now) != now)`